File: bindings/rust/src/schema.rs
//! Schema validation against the twenty-one embedded JSON Schemas.
//! The schema files are compiled into the library (include_str!), which
//! keeps the crate pure - no filesystem at run time - and therefore
//! WebAssembly-ready with zero changes.

use regex::Regex;
use serde_json::{Map, Value};
use std::collections::HashMap;
use std::sync::OnceLock;

use crate::canonical::infer_kind;

const BASE: &str = "https://causalontology.org/schema/";

/// kind -> schema file. Three token kinds keep their original 1.0.0-reserved
/// file names (individual/token/state); the id scheme is the whole word.
fn schema_file(kind: &str) -> Option<&'static str> {
    Some(match kind {
        "occurrent" => "occurrent.schema.json",
        "causal_relation_object" => "causal_relation_object.schema.json",
        "continuant" => "continuant.schema.json",
        "realizable" => "realizable.schema.json",
        "stratum" => "stratum.schema.json",
        "bridge" => "bridge.schema.json",
        "cross_stratal_seam" => "cross_stratal_seam.schema.json",
        "port" => "port.schema.json",
        "conduit" => "conduit.schema.json",
        "quality" => "quality.schema.json",
        "token_individual" => "individual.schema.json",
        "token_occurrence" => "token.schema.json",
        "state_assertion" => "state.schema.json",
        "token_causal_claim" => "token_causal_claim.schema.json",
        "attitude" => "attitude.schema.json",
        "predicted_occurrence" => "predicted_occurrence.schema.json",
        "prediction_error" => "prediction_error.schema.json",
        "assertion" => "assertion.schema.json",
        "enrichment" => "enrichment.schema.json",
        "retraction" => "retraction.schema.json",
        "succession" => "succession.schema.json",
        _ => return None,
    })
}

// The twenty-one whole-word schemas, embedded at compile time and keyed by
// their file name so cross-file $ref resolution is a simple lookup.
static SCHEMAS: OnceLock<HashMap<&'static str, Value>> = OnceLock::new();

macro_rules! embed {
    ($m:ident, $file:literal) => {
        $m.insert($file, serde_json::from_str(include_str!(
            concat!("../spec_schema/", $file))).unwrap());
    };
}

fn schemas() -> &'static HashMap<&'static str, Value> {
    SCHEMAS.get_or_init(|| {
        let mut m: HashMap<&'static str, Value> = HashMap::new();
        embed!(m, "occurrent.schema.json");
        embed!(m, "causal_relation_object.schema.json");
        embed!(m, "continuant.schema.json");
        embed!(m, "realizable.schema.json");
        embed!(m, "stratum.schema.json");
        embed!(m, "bridge.schema.json");
        embed!(m, "cross_stratal_seam.schema.json");
        embed!(m, "port.schema.json");
        embed!(m, "conduit.schema.json");
        embed!(m, "quality.schema.json");
        embed!(m, "individual.schema.json");
        embed!(m, "token.schema.json");
        embed!(m, "state.schema.json");
        embed!(m, "token_causal_claim.schema.json");
        embed!(m, "attitude.schema.json");
        embed!(m, "predicted_occurrence.schema.json");
        embed!(m, "prediction_error.schema.json");
        embed!(m, "assertion.schema.json");
        embed!(m, "enrichment.schema.json");
        embed!(m, "retraction.schema.json");
        embed!(m, "succession.schema.json");
        m
    })
}

fn schema_by_file(file: &str) -> &'static Value {
    schemas().get(file).unwrap_or_else(|| panic!("unknown schema file: {}", file))
}

fn navigate<'a>(mut node: &'a Value, pointer: &str) -> &'a Value {
    for part in pointer.split('/') {
        if part.is_empty() {
            continue;
        }
        node = node.get(part).expect("unresolvable $ref pointer");
    }
    node
}

/// Resolve local (`#/...`) and cross-file
/// (`https://causalontology.org/schema/<file>#/...`) $refs to a concrete
/// schema node and the root document it belongs to.
fn resolve<'a>(mut schema: &'a Value, mut root: &'a Value)
               -> (&'a Value, &'a Value) {
    while let Some(Value::String(r)) = schema.get("$ref") {
        if let Some(rest) = r.strip_prefix("#/") {
            schema = navigate(root, rest);
        } else if let Some(rest) = r.strip_prefix(BASE) {
            let (file, pointer) = match rest.split_once("#/") {
                Some((f, p)) => (f, Some(p)),
                None => (rest, None),
            };
            root = schema_by_file(file);
            schema = match pointer {
                Some(p) => navigate(root, p),
                None => root,
            };
        } else {
            panic!("unsupported $ref: {}", r);
        }
    }
    (schema, root)
}

fn type_matches(t: &str, value: &Value) -> bool {
    match t {
        "object" => value.is_object(),
        "array" => value.is_array(),
        "string" => value.is_string(),
        // serde keeps booleans distinct from numbers, so integer/number
        // naturally reject a JSON boolean (as RFC 8785 / the Python ref do).
        "number" => value.is_number(),
        "integer" => value.is_i64() || value.is_u64(),
        "boolean" => value.is_boolean(),
        _ => false,
    }
}

fn check(value: &Value, schema: &Value, root: &Value, path: &str,
         errors: &mut Vec<String>) {
    let (schema, root) = resolve(schema, root);

    if let Some(Value::Array(branches)) = schema.get("oneOf") {
        let mut passing = 0;
        for branch in branches {
            let mut sub = Vec::new();
            check(value, branch, root, path, &mut sub);
            if sub.is_empty() {
                passing += 1;
            }
        }
        if passing != 1 {
            errors.push(format!(
                "{}: matches {} of the oneOf branches (need exactly 1)",
                path, passing));
        }
        return;
    }

    if let Some(Value::String(t)) = schema.get("type") {
        if !type_matches(t, value) {
            errors.push(format!("{}: expected {}", path, t));
            return;
        }
    }

    if let Some(c) = schema.get("const") {
        if value != c {
            errors.push(format!("{}: must equal {}", path, c));
        }
    }
    if let Some(Value::Array(options)) = schema.get("enum") {
        if !options.contains(value) {
            errors.push(format!("{}: {} not in enumeration", path, value));
        }
    }
    if let (Some(Value::String(p)), Some(s)) =
        (schema.get("pattern"), value.as_str()) {
        let re = Regex::new(p).expect("invalid schema pattern");
        if !re.is_match(s) {
            errors.push(format!("{}: {:?} does not match {}", path, s, p));
        }
    }
    if let (Some(min), Some(s)) =
        (schema.get("minLength").and_then(Value::as_u64), value.as_str()) {
        if (s.chars().count() as u64) < min {
            errors.push(format!("{}: shorter than minLength", path));
        }
    }
    if let (Some(min), Some(x)) =
        (schema.get("minimum").and_then(Value::as_f64), value.as_f64()) {
        if x < min {
            errors.push(format!("{}: below minimum {}", path, min));
        }
    }
    if let (Some(max), Some(x)) =
        (schema.get("maximum").and_then(Value::as_f64), value.as_f64()) {
        if x > max {
            errors.push(format!("{}: above maximum {}", path, max));
        }
    }

    if let Value::Array(items) = value {
        if let Some(min) = schema.get("minItems").and_then(Value::as_u64) {
            if (items.len() as u64) < min {
                errors.push(format!("{}: fewer than {} items", path, min));
            }
        }
        if let Some(item_schema) = schema.get("items") {
            for (i, item) in items.iter().enumerate() {
                check(item, item_schema, root,
                      &format!("{}[{}]", path, i), errors);
            }
        }
    }

    if let Value::Object(map) = value {
        let empty = Map::new();
        let props = schema.get("properties")
            .and_then(Value::as_object).unwrap_or(&empty);
        if let Some(Value::Array(required)) = schema.get("required") {
            for req in required {
                if let Some(name) = req.as_str() {
                    if !map.contains_key(name) {
                        errors.push(format!(
                            "{}: required property '{}' missing", path, name));
                    }
                }
            }
        }
        if schema.get("additionalProperties") == Some(&Value::Bool(false)) {
            for key in map.keys() {
                if !props.contains_key(key) {
                    errors.push(format!(
                        "{}: additional property '{}'", path, key));
                }
            }
        }
        for (key, sub) in props {
            if let Some(v) = map.get(key) {
                check(v, sub, root, &format!("{}.{}", path, key), errors);
            }
        }
    }
}

/// (ok, reasons) - structural validity against the kind's JSON Schema.
pub fn validate_schema(obj: &Map<String, Value>, kind: Option<&str>)
                       -> (bool, Vec<String>) {
    let kind = match kind {
        Some(k) => k.to_string(),
        None => match infer_kind(obj) {
            Ok(k) => k,
            Err(e) => return (false, vec![e]),
        },
    };
    let root = match schema_file(&kind) {
        Some(file) => schema_by_file(file),
        None => return (false, vec![format!("unknown kind: {}", kind)]),
    };
    let mut errors = Vec::new();
    check(&Value::Object(obj.clone()), root, root, "$", &mut errors);
    (errors.is_empty(), errors)
}

File: bindings/r/R/schema.R
# causalontology-r -- schema.R
#
# Schema validation against spec/schema/*.schema.json, ported from the
# reference bindings/python/causalontology/schema.py.
#
# A deliberately small interpreter for exactly the JSON Schema keywords the
# seventeen Causalontology schemas use: type, const, enum, pattern,
# required, properties, additionalProperties, items, minItems, minLength,
# minimum, maximum, oneOf, local $ref (#/$defs/...), and cross-file $ref to
# a sibling schema (https://causalontology.org/schema/<file>.schema.json#/
# ...). "format" is treated as an annotation, as the 2020-12 draft does by
# default.

# kind -> schema file. Three token kinds keep their original 1.0.0-reserved
# file names (individual/token/state); the id scheme is the whole word.
co_schema_files <- c(
  occurrent  = "occurrent.schema.json",
  causal_relation_object = "causal_relation_object.schema.json",
  continuant = "continuant.schema.json",
  realizable = "realizable.schema.json",
  stratum    = "stratum.schema.json",
  bridge     = "bridge.schema.json",
  port       = "port.schema.json",
  conduit    = "conduit.schema.json",
  quality    = "quality.schema.json",
  token_individual   = "individual.schema.json",
  token_occurrence   = "token.schema.json",
  state_assertion    = "state.schema.json",
  token_causal_claim = "token_causal_claim.schema.json",
  assertion  = "assertion.schema.json",
  enrichment = "enrichment.schema.json",
  retraction = "retraction.schema.json",
  succession = "succession.schema.json"
)

co_schema_base <- "https://causalontology.org/schema/"

co_schema_dir <- function() {
  env <- Sys.getenv("CAUSALONTOLOGY_SPEC", unset = "")
  if (nzchar(env)) return(file.path(env, "schema"))
  # In-repo / sourced: the standard's spec/schema is the source of truth.
  root <- tryCatch(co_repo_root(), error = function(e) NULL)
  if (!is.null(root) && dir.exists(file.path(root, "spec", "schema"))) {
    return(file.path(root, "spec", "schema"))
  }
  # Installed package: the schemas bundled under inst/schema.
  p <- system.file("schema", package = "causalontology")
  if (nzchar(p) && dir.exists(p)) return(p)
  stop("cannot locate the specification schemas")
}

# Load and cache one schema file by its filename.
co_load_schema_file <- function(filename) {
  if (!exists("schemas", envir = co_state, inherits = FALSE)) {
    assign("schemas", list(), envir = co_state)
  }
  cache <- get("schemas", envir = co_state, inherits = FALSE)
  if (!co_has_key(cache, filename)) {
    path <- file.path(co_schema_dir(), filename)
    text <- paste(readLines(path, warn = FALSE, encoding = "UTF-8"),
                  collapse = "\n")
    cache[[filename]] <- co_parse_json(text)
    assign("schemas", cache, envir = co_state)
  }
  cache[[filename]]
}

co_load_schema <- function(kind) {
  if (!(kind %in% names(co_schema_files))) stop("unknown kind: ", kind)
  co_load_schema_file(co_schema_files[[kind]])
}

# Navigate a JSON pointer body (already stripped of a leading "#/" or "").
co_schema_navigate <- function(doc, pointer) {
  node <- doc
  for (part in strsplit(pointer, "/", fixed = TRUE)[[1]]) {
    if (!nzchar(part)) next
    node <- node[[part]]
  }
  node
}

# Follow local and cross-file $ref chains to a concrete node + its root.
# Returns list(schema = <node>, root = <root doc>).
co_schema_resolve <- function(schema, root) {
  while (co_is_obj(schema) && co_has(schema, "$ref")) {
    ref <- schema[["$ref"]]
    if (startsWith(ref, "#/")) {
      schema <- co_schema_navigate(root, substring(ref, 3L))
    } else if (startsWith(ref, co_schema_base)) {
      rest <- substring(ref, nchar(co_schema_base) + 1L)
      idx <- regexpr("#/", rest, fixed = TRUE)
      if (idx > 0L) {
        filename <- substr(rest, 1L, idx - 1L)
        pointer <- substring(rest, idx + 2L)
      } else {
        filename <- rest
        pointer <- ""
      }
      root <- co_load_schema_file(filename)
      schema <- if (nzchar(pointer)) co_schema_navigate(root, pointer) else root
    } else {
      stop("unsupported $ref: ", ref)
    }
  }
  list(schema = schema, root = root)
}

# Does this value have the given JSON Schema type?
co_schema_type_ok <- function(value, t) {
  if (identical(t, "object"))  return(co_is_obj(value))
  if (identical(t, "array"))   return(co_is_arr(value))
  if (identical(t, "string"))  return(co_is_str(value))
  if (identical(t, "boolean")) return(co_is_bool(value))
  if (identical(t, "integer")) return(co_is_num(value) &&
                                        isTRUE(as.numeric(value) ==
                                               floor(as.numeric(value))))
  if (identical(t, "number"))  return(co_is_num(value))  # excludes logicals
  stop("unsupported schema type: ", t)
}

# One recursive check; errors accumulate in the collector environment.
co_schema_check <- function(value, schema, root, path, errs) {
  rr <- co_schema_resolve(schema, root)
  schema <- rr$schema
  root <- rr$root

  if (co_has(schema, "oneOf")) {
    passing <- 0L
    branches <- schema[["oneOf"]]
    for (i in seq_along(branches)) {
      sub_errs <- new.env(parent = emptyenv())
      sub_errs$msgs <- character(0)
      co_schema_check(value, branches[[i]], root, path, sub_errs)
      if (length(sub_errs$msgs) == 0L) passing <- passing + 1L
    }
    if (passing != 1L) {
      errs$msgs <- c(errs$msgs, sprintf(
        "%s: matches %d of the oneOf branches (need exactly 1)",
        path, passing))
    }
    return(invisible(NULL))
  }

  t <- co_get(schema, "type")
  if (!is.null(t)) {
    if (!co_schema_type_ok(value, t)) {
      errs$msgs <- c(errs$msgs, sprintf("%s: expected %s", path, t))
      return(invisible(NULL))
    }
  }

  if (co_has(schema, "const") && !co_equal(value, schema[["const"]])) {
    errs$msgs <- c(errs$msgs, sprintf("%s: must equal '%s'",
                                      path, as.character(schema[["const"]])[[1]]))
  }
  if (co_has(schema, "enum")) {
    enum <- schema[["enum"]]
    hit <- FALSE
    for (i in seq_along(enum)) {
      if (co_equal(value, enum[[i]])) { hit <- TRUE; break }
    }
    if (!hit) {
      shown <- if (co_is_str(value)) value else co_jcs(value)
      errs$msgs <- c(errs$msgs, sprintf("%s: '%s' not in enumeration",
                                        path, shown))
    }
  }
  if (co_has(schema, "pattern") && co_is_str(value)) {
    if (!grepl(schema[["pattern"]], value, perl = TRUE)) {
      errs$msgs <- c(errs$msgs, sprintf("%s: '%s' does not match %s",
                                        path, value, schema[["pattern"]]))
    }
  }
  if (co_has(schema, "minLength") && co_is_str(value)) {
    if (nchar(value, type = "chars") < as.numeric(schema[["minLength"]])) {
      errs$msgs <- c(errs$msgs, sprintf("%s: shorter than minLength", path))
    }
  }
  if (co_has(schema, "minimum") && co_is_num(value)) {
    if (as.numeric(value) < as.numeric(schema[["minimum"]])) {
      errs$msgs <- c(errs$msgs, sprintf("%s: below minimum %s", path,
                                        co_jcs_number(schema[["minimum"]])))
    }
  }
  if (co_has(schema, "maximum") && co_is_num(value)) {
    if (as.numeric(value) > as.numeric(schema[["maximum"]])) {
      errs$msgs <- c(errs$msgs, sprintf("%s: above maximum %s", path,
                                        co_jcs_number(schema[["maximum"]])))
    }
  }

  if (co_is_arr(value)) {
    if (co_has(schema, "minItems") &&
        length(value) < as.numeric(schema[["minItems"]])) {
      errs$msgs <- c(errs$msgs, sprintf("%s: fewer than %.0f items", path,
                                        as.numeric(schema[["minItems"]])))
    }
    if (co_has(schema, "items")) {
      for (i in seq_along(value)) {
        co_schema_check(value[[i]], schema[["items"]], root,
                        sprintf("%s[%d]", path, i - 1L), errs)
      }
    }
  }

  if (co_is_obj(value)) {
    props <- co_get(schema, "properties", co_obj())
    prop_names <- names(props)
    if (is.null(prop_names)) prop_names <- character(0)
    for (req in co_strings(co_get(schema, "required"))) {
      if (!co_has(value, req)) {
        errs$msgs <- c(errs$msgs, sprintf(
          "%s: required property '%s' missing", path, req))
      }
    }
    if (identical(co_get(schema, "additionalProperties"), FALSE)) {
      value_names <- names(value)
      if (is.null(value_names)) value_names <- character(0)
      for (key in value_names) {
        if (!(key %in% prop_names)) {
          errs$msgs <- c(errs$msgs, sprintf(
            "%s: additional property '%s'", path, key))
        }
      }
    }
    for (key in prop_names) {
      if (co_has(value, key)) {
        co_schema_check(value[[key]], props[[key]], root,
                        sprintf("%s.%s", path, key), errs)
      }
    }
  }

  invisible(NULL)
}

# list(ok, errors) -- structural validity against the kind's JSON Schema.
co_validate_schema <- function(obj, kind = NULL) {
  if (is.null(kind)) kind <- co_infer_kind(obj)
  root <- co_load_schema(kind)
  errs <- new.env(parent = emptyenv())
  errs$msgs <- character(0)
  co_schema_check(obj, root, root, "$", errs)
  list(ok = length(errs$msgs) == 0L, errors = errs$msgs)
}
